core/src/main/java/org/apache/iceberg/CherryPickFromSnapshot.java

```diff
@@ -0,0 +1,225 @@
+ * In an audit workflow, new data is written to an orphan snapshot that is not committed as the table's
+ * current state until it is audited. After auditing a change, it may need to be applied or cherry-picked
+ * on top of the latest snapshot instead of the one that was current when the audited changes were created.
+ *
+ * This class adds support for cherry-picking the changes from an orphan snapshot by applying them to
+ * the current snapshot. The output of the operation is a new snapshot with the changes from cherry-picked
+ * snapshot.
+ *
+ * Cherry-picking should apply the exact set of changes that were done in the original commit.
+ *  - All added files should be added to the new version.
+ *  - Todo: If files were deleted, then those files must still exist in the data set.
+ *  - Does not support Overwrite operations currently. Overwrites are considered as conflicts.
+ *
+ */
+class CherryPickFromSnapshot extends MergingSnapshotProducer<AppendFiles> implements CherryPick {
+  private final TableOperations ops;
+  private TableMetadata base = null;
+  private Long cherryPickSnapshotId = null;
+  private final AtomicInteger manifestCount = new AtomicInteger(0);
+
+  CherryPickFromSnapshot(TableOperations ops) {
+    super(ops);
+    this.ops = ops;
+    this.base = ops.refresh();
+  }
+
+  @Override
+  protected AppendFiles self() {
+    throw new UnsupportedOperationException("Not supported");
+  }
+
+  /**
+   * We only cherry pick for appends right now
+   * @return
+   */
+  @Override
+  protected String operation() {
+    return DataOperations.APPEND;
+  }
+
+  @Override
+  public CherryPickFromSnapshot fromSnapshotId(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    this.cherryPickSnapshotId = snapshotId;
+    return this;
+  }
+
+  /**
+   * Apply the pending changes and return the uncommitted changes for validation.
+   * <p>
+   * This does not result in a permanent update.
+   *
+   * @return the uncommitted changes that would be committed by calling {@link #commit()}
+   * @throws ValidationException      If the pending changes cannot be applied to the current metadata
+   * @throws IllegalArgumentException If the pending changes are conflicting or invalid
+   */
+  @Override
+  public Snapshot apply() {
+    ValidationException.check(cherryPickSnapshotId != null,
+        "Cannot cherry pick unknown version: call fromSnapshotId");
+
+    Snapshot cherryPickSnapshot = base.snapshot(cherryPickSnapshotId);
+    String wapId = stagedWapId(cherryPickSnapshot);
+    ValidationException.check(wapId != null,
+        "Snapshot provided for cherrypick didn't have wap.id set on it. " +
+            "Only snapshots that were staged can be cherrypicked from.");
+    ValidationException.check(!base.isWapIdPublished(Long.parseLong(wapId)),
+        "Duplicate request to cherry pick wap id that was published already: %s", wapId);
```

Comment:
I like this check, but I think it should be whether the snapshot ID was published, not necessarily the WAP id.

Reply:
This is checking on the client provided `wap.id` which could be different from the snapshot id. It's part of our wap workflow. How we use this in our wap workflow is: we are setting an incoming batch-id (an id for a new delta of data) as the `wap.id` and use that to audit our snapshot. If the same batch (wap.id) was published previously we reject that snapshot. This way multiple writers trying to cherry-pick the same batch would be safeguarded as the second one trying to write would get this `ValidationException`. What do you think?

Reply:
I think it's a good idea to have the check, but shouldn't we use Iceberg's snapshot ID? Then we can check whether that snapshot is in the current snapshot's ancestors, or whether any ancestor was picked from that snapshot.

If we use WAP ID for this, then we have to rely on having a WAP ID, which might not be the case if we are picking a commit that was rolled back like in the example I gave in another comment: if a table has snapshots A -> B -> C and is rolled back to A, it should be possible to pick C on top of A.

Reply:
> I think it's a good idea to have the check, but shouldn't we use Iceberg's snapshot ID? Then we can check whether that snapshot is in the current snapshot's ancestors, or whether any ancestor was picked from that snapshot.

I think we are talking about two different validations. viz.
1) check if a snapshot is already in the existing chain of snapshots so that same snapshot isn't cherrypicked again later down the chain (this, I think is what you are referring to), 
2) check if the incoming snapshot has a `wap.id` that was already committed up the chain. Here the `wap.id` is a duplicate id across two or more unique snapshots that those writers are trying to cherry-pick. (this is what I'm trying to check for)


I agree that the rollback usecase is a problem when checking using wap.id. In general I think we can make `wap.id` optional on the incoming snapshot. In validations, we can perform both checks 1 & 2. I can make wap.id optional and perform check 2 only if there is a `wap.id` on the snapshot.

My point is, check 2 is essential to our usecase as it allows writers to write concurrently and protects against duplicate wap.ids from entering the system. Do you think this is something we can can incorporate into this workflow? Or do you suggest we do this another way.

Reply:
Okay, that makes sense. I was thinking that a WAP ID was unique to reach write but it makes sense that you may want to use it that way and that you would want to check WAP ID as well. Let's add both checks.

I think that this should _always_ check that the snapshot itself was not already picked into the current table state. I'm not sure whether to always check that the WAP ID was not picked into the current table state as well. I don't think that it would cause problems if we included the check by default, though. So it's probably easier for everyone if both are on by default. Sound okay?

Reply:
+1 .. if we need an override to skip the 2nd check I can open a separate issue later. But for now defaulting to both checks works just fine.

Reply:
Wanted to call out two things about the cherrypick, w.r.t to validation checks.. 
1. if  A->B->C is the current snapshot chain and we rolled-back to A. Then subsequently we  cherrypicked C over A. If we perform check 1 ( to check if C exists in the existing chain) then C will fail this check as it's in the list of snapshots just that it's not an ancestor of A. This is why I haven't added that check yet.
2. When cherrypicking C , the cherrypick operation will create a new snapshot by applying C's appends and create a new snapshot D which is then set as the current_snapshot. So each cherrypick operation creates a new snapshot. 

how does this sound to you?